stateMutators.js

```diff
@@ -0,0 +1,174 @@
+const gameState = {
+    basketPos: 50,
+    basketValue: 0,
+    score: 0,
+    livesRemaining: 3,
+    gameLevel: 1,
+    gameMode: 'running',
+    fallingObjects: [],
+}
+
+const setBasketPos = (state, value) => {
+    return {
+        ...state,
+        basketPos: value,
+    }
+};
+
+const moveBasketLeft = (state, value) => {
+    return {
+        ...state,
+    basketPos: basketPos - value,
+    }
+};
+
+const moveBasketRight = (state, value) => {
+    return {
+        ...state,
+        basketPos: basketPos + value,
+    }
+};
+
+const setBasketValue = (state, value) => {
+    return {
+        ...state,
+        basketValue: value,
+    }
+};
+
+const resetBasketValue = (state) => {
+    return {
+        ...state,
+        basketValue: 0,
+    }
+};
+
+const setScore = (state, value) => {
+    return {
+        ...state,
+        score: value,
+    }
+};
+
+const resetScore = (state) => {
+    return {
+        ...state,
+        score: 0,
+    }
+};
+
+const loseLife = (state, value) => {
+    return {
+        ...state,
+        livesRemaining: value,
+    }
+};
+
+const resetLivesRemaining = (state) => {
+    return {
+        ...state,
+        livesRemaining: 3,
+    }
+};
+
+const setGameLevel = (state, value) => {
+    return {
+        ...state,
+        gameLevel: value,
+    }
+};
+
+const resetGameLevel = (state) => {
+    return {
+        ...state,
+        gameLevel: 0,
+    }
+};
+
+const setGameMode = (state, value) => {
+    return {
+        ...state,
+        gameMode: value,
+    }
+};
+
+const addFallingObject = (state, fallingObj) => {
+    return {
+        ...state,
+        fallingObjects: [
+            ...state.fallingObjects,
+            fallingObj,
+        ]
+    }
+};
+
+const removeFallingObject = (state, fallingObjId) => {
+    return {
+        ...state,
+        fallingObjects: [
+            ...state.fallingObjects.slice[0, fallingObjId],
+            ...state.fallingObjects.slice[fallingObjId + 1],
+        ]
+    }
+};
+
+
+// Object constructor for falling object
+function FallingObject(posX, posY, numerator, denominator) {
+    this.posX = posX;
+    this.posY = posY;
+    this.numerator = numerator;
+    this.denominator = denominator;
+    this.value = numerator / denominator;
+    this.id = id;
+}
+
+const setFallingObjPosY = (fallingObj, value) => {
+    fallingObj.posY = value;
+}
+
+
+const setFallingObjNumerator = (fallingObj, value) => {
+    fallingObj.numerator = value;
+}
+
+const setFallingObjDenominator = (fallingObj, value) => {
+    fallingObj.denominator = value;
+}
+
+// Calculating Functions
+
+const calcBasketValue = (state, basketValue, fallingObjectValue) => {
+    return {
+        ...state,
+        basketValue: basketValue + fallingObjectValue,
```

Comment:
I think this function shouldn't have that second param `basketValue` as that should be derived from the first param `state` via `state.basketValue`. And this line would need updating

Reply:
I made basket an object so that it also can have it's own x position, width and height properties which will be used for the collision detection later on. So for these mutators, I have removed basketValue and replaced them with basket. Should I remove the basket param?

Reply:
I removed basket after reading your next comment

Reply:
I saw that. TBH I didn't like it at first as it makes the state more complicated, but the width and height states opens up the possibility of having different difficulties later where you might have a smaller/bigger basket. Let's go with it. Though I don't think there's a need for the basket's y position state, or do you disagree?

> Should I remove the basket param?

yeah these values are still all accessible at `state.basket.value`, etc. So still good to remove those params, as it makes things more simple (what we should always aim for)

Reply:
I included the basket's y position because it's going to need to be set at some time, even if it's just once initially and I thought it'd be best to have all of it's properties in one place.

Reply:
The basket's Y pos does need to be set at some time but, as it will likely never need to change, I'd still argue that it doesn't really belong here IMO. The setting of the basket Y pos could just be done in the HTML+CSS as it's static. State should be as minimal as possible, and only contain dynamic data that could change throughout the lifecycle of the program.

Similar to the reason we don't include things like X and Y position of the score box in state, for example. It's not anticipated to ever change in the course of the game, so it's going to be left up to the main styles code to define that. 

But it's not super important to get this all perfect. Once this gets merged, likely later on we'll realize there was some state change we were not anticipating and need to add support for. That's normal. The main point is try to identify if any values in state are actually static and thus probably shouldn't be there. So don't worry about removing this